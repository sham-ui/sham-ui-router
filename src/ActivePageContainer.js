import { $ } from 'sham-ui-macro/ref.macro';
import { Component, insert } from 'sham-ui';
import { useStorage } from './storage';

/**
 * Options for LinkTo
 * @typedef {Object} LinkToOptions
 * @property {string} path Name of page for link. Default ''
 * @property {Object} params Params of page for link. Default {}
 * @property {string} text Text for link. Default ''
 * @property {boolean} useActiveClass Use activeClass options for active page link. Default false
 * @property {string} activeClass Class name for active link. Default 'active'
 */

/**
 * Component for link to page
 * @class LinkTo
 * @classdesc
 * @property {LinkToOptions} options
 */

/**
 * Component-container for page
 * @example
 * {% import ActivePageContainer from 'sham-ui-router/active-page-container' %}
 * ...
 * <ActivePageContainer/>
 * ...
 */
function ActivePageContainer() {
    this.ref = null;

    let lastRenderedURL = null;
    let lastRenderedComponent = null;

    this.spots = [ [
        $.routerData,
        routerData => {
            const url = routerData.url;
            const activePageComponent = routerData.activePageComponent;
            if (
                url !== lastRenderedURL ||

                // Component can changed without change url
                activePageComponent !== lastRenderedComponent
            ) {

                // Remove self from parent's children map or child ref.
                if ( this.unbind ) {
                    this.unbind();
                }

                // Remove all nested views.
                let i = this.nested.length;
                while ( i-- ) {
                    this.nested[ i ].remove();
                }
            }
            if ( null === activePageComponent ) {
                return;
            }
            insert(
                this,
                this.container,
                this,
                activePageComponent,
                routerData.activePageOptions,
                this.owner,
                this.blocks
            );
            lastRenderedURL = url;
            lastRenderedComponent = activePageComponent;
        }
    ] ];
}

export default Component( useStorage( $.routerData ), ActivePageContainer );
