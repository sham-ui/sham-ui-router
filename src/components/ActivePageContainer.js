import { Component } from 'sham-ui';
import { useStorage } from '../storage';

/**
 * Options for LinkTo
 * @typedef {Object} LinkToOptions
 * @property {string} path Name of page for link. Default ''
 * @property {Object} params Params of page for link. Default {}
 * @property {string} text Text for link. Default ''
 * @property {boolean} useActiveClass Use activeClass options for active page link. Default false
 * @property {string} activeClass Class name for active link. Default 'active'
 */

/**
 * Component for link to page
 * @class LinkTo
 * @classdesc
 * @property {LinkToOptions} options
 */

/**
 * Component-container for page
 * @example
 * {% import ActivePageContainer from 'sham-ui-router/active-page-container' %}
 * ...
 * <ActivePageContainer/>
 * ...
 */
class ActivePageContainer extends Component {
    constructor() {
        super( ...arguments );
        this.lastRendererdURL = null;
        this.lastRenderedComponent = null;
        this.ref = null;
    }

    updateSpots() {
        const url = this.options.routerData.url;
        const activePageComponent = this.options.routerData.activePageComponent;
        if (
            null !== this.lastRendererdURL &&
            (
                url !== this.lastRendererdURL ||

                // Component can changed without change url
                activePageComponent !== this.lastRenderedComponent
            )
        ) {
            this._clearContainer();
        }
        if ( null === activePageComponent ) {
            return;
        }
        window.__UI__.insert(
            this,
            this.container,
            this,
            activePageComponent,
            this.options.routerData.activePageOptions,
            this.owner,
            this.blocks
        );
        this.lastRendererdURL = url;
        this.lastRenderedComponent = activePageComponent;
    }

    _clearContainer() {

        // Remove self from parent's children map or child ref.
        if ( this.unbind ) {
            this.unbind();
        }

        // Remove all nested views.
        let i = this.nested.length;
        while ( i-- ) {
            this.nested[ i ].remove();
        }
    }
}

export default useStorage( 'routerData' )( ActivePageContainer );
