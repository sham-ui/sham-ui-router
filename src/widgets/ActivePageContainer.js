import { DI } from 'sham-ui';
import ActivePageContainer from './ActivePageContainer.sht';

export default class extends ActivePageContainer {
    constructor() {
        super( ...arguments );
        DI.bind( 'widgets:active-page-container', this );
    }

    render() {
        super.render( ...arguments );
        const router = DI.resolve( 'router' );
        this.remove();
        // eslint-disable-next-line no-undef
        ShamUIView.insert(
            this,
            this.container,
            { ref: null },
            router.activePageWidget,
            router.activePageOptions
        );
    }
}
