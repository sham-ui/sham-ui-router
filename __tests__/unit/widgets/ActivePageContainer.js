import ActivePageContainer from '../../../src/widgets/ActivePageContainer';
import renderer from 'sham-ui-test-helpers';

it( 'renders correctly', () => {
    const meta = renderer( ActivePageContainer, {} );
    expect( meta.toJSON() ).toMatchSnapshot();
} );
