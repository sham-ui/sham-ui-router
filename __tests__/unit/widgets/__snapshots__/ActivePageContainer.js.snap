// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correct process dynamic placet 1`] = `
Object {
  "Constructor": "DynamicPlaceActivePageContainer",
  "Options": Object {
    "ID": "widget",
    "containerSelector": "body",
    "types": Array [],
    "withWrapper": false,
  },
  "html": "
<h1>Title</h1>
<div>Content for dummy widget</div>
<!--ActivePageContainer-->
<!--if-->
",
}
`;

exports[`correct process dynamic placet 2`] = `
Object {
  "Constructor": "DynamicPlaceActivePageContainer",
  "Options": Object {
    "ID": "widget",
    "containerSelector": "body",
    "types": Array [],
    "withWrapper": true,
  },
  "html": "
<div class=\\"wrapper\\">
  <h1>Title</h1>
  <div>Content for dummy widget</div>
</div>
<!--if-->
",
}
`;

exports[`renders correctly 1`] = `
Object {
  "Constructor": "ActivePageContainer",
  "Options": Object {
    "ID": "widget",
    "containerSelector": "body",
    "types": Array [
      "active-page-container",
    ],
  },
  "html": "
<h1>Title</h1>
<div>Content for dummy widget</div>
",
}
`;
